// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package pdpi;

import "p4/config/v1/p4info.proto";
import "p4/v1/p4runtime.proto";
import "p4_pdpi/ir.proto";

message Tests {
  repeated Test tests = 1;
}

message Test {
  string name = 1;
  oneof kind {
    InfoTest info_test = 2;
    TableEntryTest table_entry_test = 3;
    PacketIoTest packet_io_test = 4;
  }
}

// A P4 info file, either inline or referring to one by path to allow reuse.
message P4Info {
  oneof source {
    p4.config.v1.P4Info direct = 1;
    // Path to the P4 info file.
    string indirect = 2;
  }
}

message InfoTest {
  P4Info p4info = 1;
}

message TableEntryTest {
  P4Info p4info = 1;
  repeated PiTableEntryCase pi_table_entry_cases = 2;
  repeated IrTableEntryCase ir_table_entry_cases = 3;
}

message PiTableEntryCase {
  string name = 1;
  p4.v1.TableEntry pi = 2;
}

message IrTableEntryCase {
  string name = 1;
  IrTableEntry ir = 2;
}

message PacketIoTest {
  P4Info p4info = 1;
  repeated PiPacketInCase pi_packet_in_cases = 2;
  repeated IrPacketInCase ir_packet_in_cases = 3;
}

message PiPacketInCase {
  string name = 1;
  p4.v1.PacketIn pi = 2;
}

message IrPacketInCase {
  string name = 1;
  IrPacketIn ir = 2;
}
